### PR TITLE
[feat] Added AVX2 SHA support

### PR DESCRIPTION
**Pull request by edtubbs**
Added Intel-ipsec-mb to depends
Added AVX2 SHA1, SHA256, and SHA512
Added arch and os guard to legacy code
Added nasm package to CI environment

### CODE REVIEW

**Comment by ReverseControl** [comment]
@edtubbs Would it not be best to check for AVX512 first, and then AVX2, and use the implementation in that order of precedence?

**Comment by edtubbs** [comment]
@ReverseControl thanks for the interest.  There's plans to add support for more intrinsics incrementally; AVX2 is the first.

**Comment by ReverseControl** [comment]
@edtubbs I do want to point out that for the AVX512, after your work, what needs to be done is to add another macro for AVX512 wherever you placed an AVX2 macro to run the sha function, so literally 3 places, so another 6 lines of code is all that is needed.

**Comment by edtubbs** [comment]
Thanks @ReverseControl, and there's plans to add ARM intrinsics soon as well!

**Comment by edtubbs** [comment]
@patricklodder gitian-builder https://cirrus-ci.com/github/micaelmalta/gitian-builder-dogecoin/build_avx2

**Comment by patricklodder** [comment]
@edtubbs In early commits it looks like you built `nasm` from source as part of the build process (which is much like what the depends `*_native` packages do) but later took the ubuntu package. What made you change that?

**Review by dgs2019 — APPROVED** [review]

File: depends/patches/intel-ipsec-mb/remove_digest_init.patch
Comment (chromatic) [inline depends/patches/intel-ipsec-mb/remove_digest_init.patch:10]: What's the source and purpose of this patch? The commit might be clearer if we can tell why it's here and under what circumstances we might want to remove the patch (if the library changes, for example).

Comment (edtubbs) [reply depends/patches/intel-ipsec-mb/remove_digest_init.patch:10]: Thanks, here's the related discussion.. https://github.com/intel/intel-ipsec-mb/issues/86

**Comment by edtubbs** [comment]
@patricklodder gitian-builder: https://cirrus-ci.com/build/6220865810464768

**Comment by ReverseControl** [comment]
How does #2315 affect this? @rnicoll

**Comment by patricklodder** [comment]
> How does #2315 affect this?

It doesn't really - #1438 is that same PR done for 1.14, so that code has already been used by a super-majority since November 2020.

**Comment by ReverseControl** [comment]
@patricklodder Got it. Is that issue going to be closed?

**Comment by patricklodder** [comment]
> Is that issue going to be closed?

You mean the #2315? I'd LOVE to have actual public statement so that other people than myself understand why this doesn't matter anymore, or for someone to argue against my objection and prove me wrong (other than, "it's already enforced with 1.14 and the v4 soft fork made this a fact without anyone knowing it" - that one I can write on my retraction comment)

**Comment by rnicoll** [comment]
I get:

```
make: Entering directory `/home/ubuntu/build/dogecoin/distsrc-x86_64-linux-gnu/src'
Checking glibc back compat...
dogecoin-cli: symbol clock_gettime from unsupported version GLIBC_2.17
dogecoin-tx: symbol clock_gettime from unsupported version GLIBC_2.17
make: *** [check-symbols] Error 1
make: Leaving directory `/home/ubuntu/build/dogecoin/distsrc-x86_64-linux-gnu/src'
```

while attempting a Gitian build. Anyone else seen that?

**Comment by edtubbs** [comment]
> I get:
> 
> ```
> make: Entering directory `/home/ubuntu/build/dogecoin/distsrc-x86_64-linux-gnu/src'
> Checking glibc back compat...
> dogecoin-cli: symbol clock_gettime from unsupported version GLIBC_2.17
> dogecoin-tx: symbol clock_gettime from unsupported version GLIBC_2.17
> make: *** [check-symbols] Error 1
> make: Leaving directory `/home/ubuntu/build/dogecoin/distsrc-x86_64-linux-gnu/src'
> ```
> 
> while attempting a Gitian build. Anyone else seen that?

Same here, I should've looked closer.. https://cirrus-ci.com/build/6220865810464768

Checking glibc back compat...
dogecoin-cli: symbol clock_gettime from unsupported version GLIBC_2.17
dogecoin-tx: symbol clock_gettime from unsupported version GLIBC_2.17

File: configure.ac
Comment (patricklodder) [inline configure.ac:794]: Question: since this adds a mandatory dependency for 64-bit linux, and there are issues with glibc compatibility for 1.14, perhaps this is better done on 1.21?

Comment (edtubbs) [reply configure.ac:794]: @patricklodder perhaps you're right, I mean I'd really like this to work for 14.5, but yeah glibc issues shouldn't hold it up.. @michilumin and @rnicoll, any thoughts?

Comment (patricklodder) [reply configure.ac:794]: Alternative 1: you can change the autoconf routine to set an enable flag if it's found, and make the configure-time dependency optional. When built with depends, it will then detect and enable for x86_64 linux, but not for the others? 

Alternative 2: you can also do a build flag, forget about depends and the release for a bit, and we can get some experience by running this in the field off a v1.14.5 tag built for-purpose. I'd totally put that on a few nodes to gather data on the performance if you'd like.

In both cases I'd need this to be reviewed thoroughly because this introduces a completely new dependency. @michilumin How far along is the actual review? If this is likely to be fully reviewed soon, and the I can drop some things and help fixing the glibc issue. I'm not going to do that if there's no in-depth review though.

Comment (edtubbs) [reply configure.ac:794]: I can add a build flag for initial support, and I'm definitely interested in the peformance data you can gather.

**Comment by patricklodder** [comment]
@edtubbs I will spend some time to do a full review and test (already know that I will have findings, but at a glance this will be mostly on the build system integration) to push this forward a bit. Let me know if you'd prefer me to PR you my recommendations on your branch, or if you would like me to just comment inline.

**Comment by edtubbs** [comment]
> @edtubbs I will spend some time to do a full review and test (already know that I will have findings, but at a glance this will be mostly on the build system integration) to push this forward a bit. Let me know if you'd prefer me to PR you my recommendations on your branch, or if you would like me to just comment inline.

Thanks!  PR to my branch works for me.

**Comment by patricklodder** [comment]
@edtubbs I have opened edtubbs#2 to improve the build system.

These changes:
- Make `nasm` to be a native dependency, because that's what it is (built for the build system, not for the target host) This will save a lot of headaches later.
- Add an arch/os guard to the depends system because it makes no sense to compile depends for systems that we later filter out for building in our autoconf scripts
- Since the build flag on autoconf defaults to "no", add a build flag for the depends system `AVX2=1` that enables the dependencies to, similarly defaulting to "no". This takes care of not having to worry about this with the gitian building and release binaries, yet.

All this, together with the build flag, effectively makes AVX2 an experimental feature, so I added an experimental build to the CI, making sure that this actually sees tests and any issues get flagged. Other optimizations can be added as an experimental feature too. We do need to document this, but I can do that outside of this PR.

As for merge strategy, if you agree to merge that PR in, I'd recommend to do a rebase on 1.14.5-dev, squash everything from your first commit up to and including my first commit that changes the package descriptors, but keep my last 2 commits intact for resp. the depends build flag and the experimental CI build, for a cleaner future reference.

**Comment by patricklodder** [comment]
Regarding testing: I am unable to reproduce the improvements advertised in #2104 with this on my build box, but that has virtualization and AMD CPUs (that do claim to support avx2 though), so I guess this mostly means that this needs more testing, underlining the case for making the feature experimental:

```
vendor_id	: AuthenticAMD
cpu family	: 23
model		: 1
model name	: AMD EPYC 7601 32-Core Processor
stepping	: 2
microcode	: 0x1000065
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm rep_good nopl cpuid extd_apicid tsc_known_freq pni pclmulqdq ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw perfctr_core ssbd ibpb vmmcall fsgsbase tsc_adjust bmi1 avx2 smep bmi2 rdseed adx smap clflushopt sha_ni xsaveopt xsavec xgetbv1 virt_ssbd arat arch_capabilities
```

Here are my results:

##### original

```
$time test_dogecoin 
Running 245 test cases...

*** No errors detected

real	1m17.347s
user	1m13.488s
sys	0m16.263s
```

##### with AVX2

```
$time test_dogecoin-avx2
Running 245 test cases...

*** No errors detected

real	1m26.864s
user	1m22.626s
sys	0m15.586s
```

**Comment by edtubbs** [comment]
> @edtubbs I have opened [edtubbs#2](https://github.com/edtubbs/dogecoin/pull/2) to improve the build system.
> 
> These changes:
> 
> * Make `nasm` to be a native dependency, because that's what it is (built for the build system, not for the target host) This will save a lot of headaches later.
> * Add an arch/os guard to the depends system because it makes no sense to compile depends for systems that we later filter out for building in our autoconf scripts
> * Since the build flag on autoconf defaults to "no", add a build flag for the depends system `AVX2=1` that enables the dependencies to, similarly defaulting to "no". This takes care of not having to worry about this with the gitian building and release binaries, yet.
> 
> All this, together with the build flag, effectively makes AVX2 an experimental feature, so I added an experimental build to the CI, making sure that this actually sees tests and any issues get flagged. Other optimizations can be added as an experimental feature too. We do need to document this, but I can do that outside of this PR.
> 
> As for merge strategy, if you agree to merge that PR in, I'd recommend to do a rebase on 1.14.5-dev, squash everything from your first commit up to and including my first commit that changes the package descriptors, but keep my last 2 commits intact for resp. the depends build flag and the experimental CI build, for a cleaner future reference.

Agreed nasm should be native. I had left it and the intel lib non native to support Win64 in the future, there’s currently issues building with mingw.  A host guard in source builds the legacy code… https://github.com/dogecoin/dogecoin/pull/2491/commits/d5d9b709258582b9e8fa81f9f2f8d760a5033c9c

**Comment by patricklodder** [comment]
> I had left it and the intel lib non native to support Win64 in the future, issue there where nasm is assembling the intel lib with the wrong calling convention.

The reason for `nasm` to stay non-native would be if we'd need to compile it to run on Windows, but I'm quite sure we only need it on the build host, and only to compile `intel-ipsec-mb`, and then it can be discarded, we don't even need it for compilation of the Dogecoin Core sources with the `--with-intel-avx2` flag, it's really only for compiling the dependency. Correct me if I'm wrong please.

While we're on that subject, I saw that `nasm` supports ELF and PE, but I don't see anything about Mach-O, which is needed for macOS. I also wasn't able to (easily) find documentation if there is any way for `intel-ipsec-mb` to be (cross-)compiled for macOS. This is worth thinking about, because the Intel macOS release is the second most downloaded binary, after Win64. Looking at `intel-ipsec-mb`'s capabilities right now, and Apple actively switching away from Intel hardware, I'm not sure if this will ever be supported and/or easy to hack into our cross-compilation process... I wouldn't make that a showstopper for the feature, but worth noting - how do you see this?

As for the "incubation phase" of this feature, I'd propose that we keep this in the experimental build (i.e. not part of the released binaries) until we have gathered (crowd sourced?) more data about the effect performance under different CPU models AND have at the very least Win64 integration working. After we have that, we can proceed to switch the default-off to a default-on and build it into binary releases. Reasonable?

**Comment by edtubbs** [comment]
> > I had left it and the intel lib non native to support Win64 in the future, issue there where nasm is assembling the intel lib with the wrong calling convention.
> 
> The reason for `nasm` to stay non-native would be if we'd need to compile it to run on Windows, but I'm quite sure we only need it on the build host, and only to compile `intel-ipsec-mb`, and then it can be discarded, we don't even need it for compilation of the Dogecoin Core sources with the `--with-intel-avx2` flag, it's really only for compiling the dependency. Correct me if I'm wrong please.
> 
> While we're on that subject, I saw that `nasm` supports ELF and PE, but I don't see anything about Mach-O, which is needed for macOS. I also wasn't able to (easily) find documentation if there is any way for `intel-ipsec-mb` to be (cross-)compiled for macOS. This is worth thinking about, because the Intel macOS release is the second most downloaded binary, after Win64. Looking at `intel-ipsec-mb`'s capabilities right now, and Apple actively switching away from Intel hardware, I'm not sure if this will ever be supported and/or easy to hack into our cross-compilation process... I wouldn't make that a showstopper for the feature, but worth noting - how do you see this?
> 
> As for the "incubation phase" of this feature, I'd propose that we keep this in the experimental build (i.e. not part of the released binaries) until we have gathered (crowd sourced?) more data about the effective performance under different CPU models AND have at the very least Win64 integration working. After we have that, we can proceed to switch the default-off to a default-on and build it into binary releases. Reasonable?

Agreed, we'll need build variability in depends to compile nasm (temporarily) for only those x86 hosts building the intel code, including Linux and Win64 this go around.  For macOS, SHA intrinsics will likely be ARM-based and the next part of the broader intrinsics effort in https://github.com/dogecoin/dogecoin/issues/2104#issuecomment-912108654.  I definitely like the idea of sourcing more data collectively with an experimental build of core.  We're taking performance metrics from crypto_tests that validate the implementation today, perhaps the duration of rpc tests can be analyzed or a new one written?

**Review by dgs2019 — DISMISSED** [review]

**Comment by patricklodder** [comment]
> for only those x86 hosts building the intel code, including Linux and Win64 this go around

edtubbs#2 doesn't exclude that, just need to add a line to packages.mk that enables the dependency on win64 builds once that works: `avx2_x86_64_w64_packages:=intel-ipsec-mb` and then that's that. I think that at this time I'd recommend to do that in a second, independent PR, to not further threaten the merge of this PR.

> I definitely like the idea of sourcing more data collectively with an experimental build of core. We're taking performance metrics from crypto_tests that validate the implementation today, perhaps the duration of rpc tests can be analyzed or a new one written?

there's `src/bench` too, which would be the ideal place to run performance related benchmarks, but I think it's been broken since 1.14.0 (i.e. the code has simply not been fully ported) and I haven't taken the time to fix it yet. We can target fixing it and including some relevant benchmarks for 1.14.6 though...

**Review by patricklodder — COMMENTED** [review]
tested ACK on 7dae118:

- code review ACK
- tested that this does not influence gitian builds
- tested that `AVX2=1` and `--with-intel-avx2` flags work

I still have 2 non-blocking concerns:
1. I'm not 100% sure about `intel-ipsec-mb`, this needs to be fully vetted for inclusion in release binaries. For now, I have over the course of a month browsed the [source repo]( https://github.com/intel/intel-ipsec-mb) on multiple occasions, looking for funny stuff in source and PRs, haven't found anything of deep concern.
2. There are some references to a larger "intrinsics effort" in comments - would be good to see that written out so that we all get a good feeling for where we're headed.

**IMPORTANT: Needs a second review, as I contributed code to this.**

**Comment by edtubbs** [comment]
> 2. There are some references to a larger "intrinsics effort" in comments - would be good to see that written out so that we all get a good feeling for where we're headed.

I entered a comment, kept it brief for the moment, but we can discuss more details here…
https://github.com/dogecoin/dogecoin/issues/2104#issuecomment-932993036

**Review by michilumin — APPROVED** [review]
Code review: looks OK, and sensible from the intrinsics side of things.  Will be trying to test this for perf later tonight but I'm not sure if a perf gain is necessarily a go/no go on this as it can be refined in later versions even if it remains experimental for a bit.  "Audit" of intel-ipsec-mb should be ongoing as we refine this feature in future versions, and should gate it being flagged as non-experimental.

**Comment by patricklodder** [comment]
Thanks!

> I'm not sure if a perf gain is necessarily a go/no go on this as it can be refined in later versions even if it remains experimental for a bit.

Agreed. It doesn't matter as long as it's experimental and not part of the released binaries.

**Pull request merged**
